modules/invibesBidAdapter.js

```diff
@@ -0,0 +1,438 @@
+  }
+
+  invibes.bidResponse = responseObj;
+
+  let ads = responseObj.Ads;
+
+  if (!Array.isArray(ads) || ads.length < 1) {
+    if (responseObj.AdReason != null) {
+      utils.logInfo('Invibes Adapter - ' + responseObj.AdReason);
+    }
+
+    utils.logInfo('Invibes Adapter - No ads available');
+    return [];
+  }
+
+  let ad = ads[0];
+
+  if (bidModel.PlacementId == null) {
+    utils.logInfo('Invibes Adapter - No Placement Id in response');
+    return [];
+  }
+
+  const bidResponses = [];
+  for (var i = 0; i < bidRequests.length; i++) {
+    var bidRequest = bidRequests[i];
+
+    if (bidModel.PlacementId === bidRequest.params.placementId) {
+      var size = getBiggerSize(bidRequest.sizes);
+
+      bidResponses.push({
+        requestId: bidRequest.bidId,
+        cpm: ad.BidPrice,
+        width: bidModel.Width || size[0],
+        height: bidModel.Height || size[1],
+        creativeId: ad.VideoExposedId,
+        currency: bidModel.Currency || CONSTANTS.DEFAULT_CURRENCY,
+        netRevenue: true,
+        ttl: CONSTANTS.TIME_TO_LIVE,
+        ad: renderCreative(bidModel),
+
+        placementCode: bidRequest.placementCode
+      });
+
+      const now = Date.now();
+      invibes.ivLogger = invibes.ivLogger || initLogger();
+      invibes.ivLogger.info('Bid auction started at ' + bidModel.AuctionStartTime + ' . Invibes registered the bid at ' + now + ' ; bid request took a total of ' + (now - bidModel.AuctionStartTime) + ' ms.');
+    } else {
+      utils.logInfo('Invibes Adapter - Incorrect Placement Id: ' + bidRequest.params.placementId);
+    }
+  }
+
+  return bidResponses;
+}
+
+function generateRandomId() {
+  return (Math.round(Math.random() * 1e12)).toString(36).substring(0, 10);
+}
+
+function getDocumentLocation(topWin) {
+  return topWin.location.href.substring(0, 300).split(/[?#]/)[0];
+}
+
+function parseQueryStringParams() {
+  var params = {};
+  try { params = JSON.parse(localStorage.ivbs); } catch (e) { }
+  var re = /[\\?&]([^=]+)=([^\\?&#]+)/g;
+  var m;
+  while ((m = re.exec(window.location.href)) != null) {
+    if (m.index === re.lastIndex) {
+      re.lastIndex++;
+    }
+    params[m[1].toLowerCase()] = m[2];
+  }
+  return params;
+}
+
+function getBiggerSize(array) {
+  var result = [0, 0];
+  for (var i = 0; i < array.length; i++) {
+    if (array[i][0] * array[i][1] > result[0] * result[1]) {
+      result = array[i];
+    }
+  }
+  return result;
+}
+
+function getTopMostWindow() {
+  var res = window;
+
+  try {
+    while (top !== res) {
+      if (res.parent.location.href.length) { res = res.parent; }
+    }
+  } catch (e) { }
+
+  return res;
+}
+
+function renderCreative(bidModel) {
+  return `<html>
+        <head><script type='text/javascript'>inDapIF=true;</script></head>
+          <body style='margin : 0; padding: 0;'>
+          creativeHtml
+          </body>
+        </html>`
+    .replace('creativeHtml', bidModel.CreativeHtml);
+}
+
+function getCappedCampaignsAsString() {
```

Comment:
What is the purpose of this function ?

Reply:
We have a campaign capping functionality that limits the amount of times the user should get an ad from the respective campaign, in a given time frame. An id is stored in localStorage, together with a view count and a timestamp. 
That function returns a formatted parameter that we send from the adapter to our system to know if the campaign can be displayed again at the moment, to the current user.